### PR TITLE
lib/model: Do Revert/Override synchronously

### PR DESCRIPTION
**Pull request by imsodin**
Writing a test for https://github.com/syncthing/syncthing/pull/6433 I realised `Override` and `Revert` happen asynchronously, i.e. you can't depend on any order of folder operations in conjunction with those. In addition they both set folder state and update the db, so as far as I see they mustn't run asynchronously.

I added a general purpose `syncChan` - I could also add two special purpose chans for overriding and reverting instead if this is "too magic".

### CODE REVIEW

**Comment by calmh** [comment]
Would it make sense to have these operations be called from the folder loop, and use the same mechanism as we do for requesting a scan for example?

**Comment by imsodin** [comment]
As these are methods on the folder "subtypes", we can't just call `f.override` from `folder`, but we'd have to register a interface for both `Revert` and `Override` like `puller`, which feels like a worse solution than this.

**Comment by calmh** [comment]
The methods could be part of a (more general) `puller` interface. Having a single interface for the three low level implementation methods seems fine, and in the loop we could handle it actor-style with a `chan func()` that the exported methods use to run things synchronously on the loop...

**Comment by imsodin** [comment]
With actor-style an interface isn't even necessary - does this look more like what you had in mind?

File: lib/model/folder.go [hunk 1]
```diff
@@ -172,9 +174,9 @@ func (f *folder) serve(ctx context.Context) {
 			l.Debugln(f, "Scanning due to timer")
 			f.scanTimerFired()
 
-		case req := <-f.scanNow:
-			l.Debugln(f, "Scanning due to request")
-			req.err <- f.scanSubdirs(req.subdirs)
+		case req := <-f.doInSyncChan:
+			l.Debugln(f, "Running something due to request")
```
Comment (calmh) [inline lib/model/folder.go]: :D

File: lib/model/folder.go [hunk 2]
```diff
@@ -224,13 +226,19 @@ func (f *folder) Jobs(_, _ int) ([]string, []string, int) {
...
-		err:     make(chan error),
+	return f.doInSync(func() error { return f.scanSubdirs(subdirs) })
+}
+
+// doInSync allows to run functions synchronously in folder.serve from exported,
+// asynchronously called methods.
+func (f *folder) doInSync(fn func() error) error {
+	req := syncRequest{
+		fn:  fn,
+		err: make(chan error),
```
Comment (calmh) [inline lib/model/folder.go]:
```suggestion
		err: make(chan error, 1),
```

lest we deadlock the loop if we take the ctx.Done() branch below

**Comment by calmh** [comment]
I guess it's a matter of taste if it's cleaner to have `folder` reach into each implementation via an interface, or have the implementation reach into the sync loop guts of `folder`... This way is fine to me as well

**Pull request merged**